WIP: parallel task runtime

kpamnany opened the pull request:
This replaces the existing fork-join threading infrastructure with a parallel task runtime (partr) that implements [parallel depth first scheduling](http://www.cs.cmu.edu/%7Echensm/papers/ConstructiveCacheSharing_spaa07.pdf). This model fully supports nested parallelism.

The default remains the original threading code. Enable partr by setting `JULIA_PARTR := 1` in your `Make.user`.

The core idea is simple -- Julia tasks can now be run by any thread. The task scheduler attempts to order task execution depth-first for provably better cache efficiency, and for true nested parallelism.

However, as tasks are an existing thing in Julia and used in a number of places, we're first introducing the infrastructure that will enable parallel tasks with this PR, keeping (hopefully) the serial semantics of the existing task interface. This PR does not introduce any new interface calls for parallel tasks -- those will be in future PRs.

All test-cases pass with `JULIA_PARTR` off (as they should). With `JULIA_PARTR` on, all test cases are currently passing on Linux and OS-X.

Cc: @JeffBezanson, @vtjnash, @yuyichao, @ViralBShah, @vchuravy, @anton-malakhov.

Commit:
threading: Integrating partr (almost done!)

Added partr code. Abstracted interface to threading infrastructure.

jtravs commented:
Any chance you could give a very simple example of what the interface would like like in user code?

kpamnany commented:
The Julia interface is not designed yet. While elaborations are possible, the essence of the interface is similar to [Cilk](https://www.cilkplus.org) so something like:

```
t1 = @spawn foo(1, 2) # foo(1, 2) will run asynchronously, possibly in another thread
res1 = @sync t1 # res1 will get the return value of foo(1, 2)
t2 = @parfor (+) i = 1:10 # iterations may run in parallel
  i-1
end
res2 = @sync t2 # res2 = 45
```

tknopp commented:
@kpamnany: Does this mean I can have a background thread/task running parallel to the foreground thread. In other words: Can I run a thread asynchronously?

kpamnany commented:
@tknopp: yes, that's what spawn will do. If you spawn a task and there's more than one thread, it will start running right away. It will continue to run until a yield point (another spawn, a sync, a parfor, or an explicit yield).

Sacha0 commented on src/forkjoin-ti.c:
```diff
@@ -0,0 +1,196 @@
+    while (jl_atomic_load_acquire(&tiarg->state) == TI_THREAD_INIT)
+        jl_cpu_pause();
+
+    // Assuming the functions called below doesn't contain unprotected GC
```
"doesn't" -> "don't"?

Sacha0 commented on src/julia_internal.h:
```diff
@@ -711,6 +710,45 @@ STATIC_INLINE char *jl_copy_str(char **to, const char *from)
+STATIC_INLINE uint64_t cong(uint64_t max, uint64_t unbias, uint64_t *seed)
+{
+    while ((*seed = 69069 * (*seed) + 362437) > unbias)
+        ;
```
Semicolon on separate line as a ghost of the empty loop body or unintentional extra space?

Sacha0 commented on src/partr.c:
```diff
@@ -0,0 +1,540 @@
+
+    init_started_thread();
+
+    // Assuming the functions called below doesn't contain unprotected GC
```
"doesn't" -> "don't" here as well?

ViralBShah commented:
Seems worthwhile to experiment with the Projects feature for this one.

davidanthoff commented:
Just out of curiosity, is this something that might make it into 1.0?

ViralBShah commented:
We will try to get it into 1.0 if it is ready before feature freeze, but not hold 1.0. I am personally hopeful that it will be ready by 1.0. Hope that helps.

StefanKarpinski commented:
If it can't make it into 1.0 in complete form, we can at least include it in experimental form and try our damndest to leave from for it in the 1.x series – we really don't want to have to wait until 2.0 for full-on threading support.

amitmurthy commented:
@kpamnany, will the spawned tasks be able to perform asynchronous IO via the libuv event loop ? Is the plan to have the main Julia thread run the event loop and perform all compute only tasks in separate threads?

Currently, in the `Distributed` module incoming requests are executed in separate tasks, and each invocation can also make additional remote calls. This leads to any IO being blocked when a worker is busy on a compute. Will it be possible for incoming requests to be executed via a threadpool and IO calls if any to be internally routed to the main Julia thread running the event loop?

kpamnany commented:
@amitmurthy: who runs the event loop (and how) is a good question. As a general statement, irrespective of Julia, unless you reserve a thread for I/O, it is possible for requests to be serviced late/very slowly. But you don't always have/want a thread to reserve. Ideally, this should be a program choice.

Executing tasks are not preempted. The API entry points (_spawn_, _sync_, and _parfor_) may cause the calling task to yield. However, this runtime allows for _sticky_ tasks, i.e. tasks that only run on the thread that started them. Sticky tasks do not yield in _spawn_ and _parfor_. So, you can create a sticky I/O task and drive the event loop from it. It's pretty straightforward to allow tasks to perform asynchronous I/O requests, but it isn't obvious how to get completion notifications. I'm not entirely sure how to do this right now but @vtjnash and @JeffBezanson have probably thought this out in greater detail (they suggested sticky tasks).

Clearly it would be a useful enhancement to this runtime to add the ability to trigger a task based on an event. But that gets us into having to define events, and decide semantics for event mux/demux and that opens many questions -- are there system events? Can multiple tasks be triggered by the same event? How about the conjunction or disjunction of multiple events? Not sure we should go down this rabbit hole right now.

JeffBezanson commented:
Our existing Tasks can already be triggered by events, so we're already in the rabbit hole. We can't fully leave this up to applications; we need to make some default choice for people.

amitmurthy commented:
Googled a bit on integrating libuv and multithreading. See
http://docs.libuv.org/en/v1.x/threading.html, http://docs.libuv.org/en/v1.x/async.html
and https://nikhilm.github.io/uvbook/threads.html#threads 

I would like to try out the following simple model in parallel to the work being done
in this PR.

1. The main Julia thread continues to handle all I/O and the event infrastructure.
2. Provide an API to spawn a Julia 0-arg function from a thread selected from a threadpool. Lets call this a compute thread.
3. The compute thread forwards all I/O and event handling calls (`sleep`, Timers, `notify`, etc.) to the main Julia thread.
4. I/O request forwarding from compute_thread -> main_thread is done via a multiple-writer, single-reader queue and `uv_async_send` to notify the event loop. All compute threads push their I/O requests onto this queue which is processed by the main_thread (running the event loop)
5. compute_threads are notified of I/O completion events via a regular system condition variable (`uv_cond_t` and `uv_cond_signal`). One condition variable per compute_thread.
6. Julia code running on a compute_thread is therefore not a Julia Task in the regular sense. It is just Julia code running in a separate thread. All calls requiring libuv facilities are forwarded to the main_thread and the compute_thread waits for its completion (on a system condition variable).
7. The fact that the Julia I/O API is designed to be a blocking interface (while being fully event driven and asynchronous under the hood) makes this model much easier to implement.

At the very least it will help in getting a handle on libuv event loop integration in a multi-threaded environment.

JeffBezanson commented:
Thanks @amitmurthy that sounds basically good. I suspect this can work with normal Tasks, though. When a Task (running on any thread) wants to do I/O, it queues its request and yields. When the I/O completes, the requesting Task can be restarted as usual.

JeffBezanson commented on src/julia_internal.h:
```diff
@@ -711,6 +710,45 @@ STATIC_INLINE char *jl_copy_str(char **to, const char *from)
+int last_arriver(arriver_t *, int);
+void *reduce(arriver_t *, reducer_t *, void *(*rf)(void *, void *), void *, int);
+#endif // JULIA_ENABLE_PARTR
+
```
All of these should maybe be in a scheduler.h; they probably shouldn't be called by miscellaneous run time system code.

JeffBezanson commented:
@kpamnany `jl_switchto(task)` should work for starting and resuming a task. Fortunately we are already storing per-thread stack address information, so if all tasks are put in sticky queues this might work now.

rveltz commented:
Any idea to make it to a package first?

anton-malakhov commented:
Hi folks! #iamintel here to help Kiran to push multi-threading forward as he's transitioned to other projects. He offered me to work on libuv-related stuff while he's finishing some other parts.

@amitmurthy, are you working on the approach you suggested on Jul 19th?
It looks good enough though it is still vulnerable to the situation when main thread gets blocked in compute-intensive user code thus all the I/O and events stuck during this time. Moving event loop into a separate dedicated I/O thread would work but it'd introduce additional overheads to single-threaded case if it always packs and sends requests to the other threads.
Studying go-lang runtime, I'd like to follow their flexibility in scheduling event polling to any threads available. Unfortunately, it is not possible with current libuv implementation but it is still possible to run multiple loops in parallel as they recommend for multi-threading. libuv can also evolve to support multithreading better and as result, an implementation explicitly communicating to a single uv_loop would require deep refactoring again. Thus, I'd suggest to start with uv_loop per thread and, as the next step, implement 'loop stealing/syncing' mechanism which would allow a request to migrate from an originating thread to another one by stealing or mailing the whole uv_loop instance or the underneath handles. Do you see any issue with this design?
Of course, if you already implemented your idea, I can work on something more important for the next release. We can also continue on Slack's #parallel

ViralBShah commented:
@amitmurthy is off grid for a the next couple of weeks.

tknopp commented:
@anton-malakhov: Would be super great to see progress here. Without this, asynchronous programming in GUI development (https://github.com/JuliaGraphics/Gtk.jl) is basically impossible.

StefanKarpinski commented:
> Without this, asynchronous programming in GUI development (https://github.com/JuliaGraphics/Gtk.jl) is basically impossible.

Isn't this mostly a problem with specific C libraries assuming that they can safely block the process? There's always the hacky `@threadcall` facility that I added some time ago. There's another crazy idea that I've floated from time to time: use `LD_PRELOAD` to replace all blocking `libc` calls with non-blocking versions that yield to the Julia scheduler.

tknopp commented:
Sure its specific C libraries, but Gtk is a pretty important thing for Julia since it is currently the only full featured way to develop desktop UIs.

If I understand it correctly Gtk has a rendering thread and regularly calls callbacks that can be implemented in Julia. If I do any computation there, the UI hangs. What one would usually do then is
- spawn a thread that does the computation
- trigger a progress bar to update the UI and inform the user that computation is done
- if thread is finished -> finish progress bar.

All this is not specific to Gtk but is the same way done in any UI framework. I am not really sure if `@threadcall` can help here. If it does would be great to here how.

amitmurthy commented:
@anton-malakhov golang's design is interesting. They appear to treat file io and socket io differently. 

From https://dave.cheney.net/2015/08/08/performance-without-the-event-loop
```
Go’s answer to c10k

In Go, syscalls are usually blocking operations, this includes reading and writing to file
descriptors. The Go scheduler handles this by finding a free thread or spawning another
to continue to service goroutines while the original thread blocks. In practice this works
well for file IO as a small number of blocking threads can quickly exhaust your local IO
bandwidth.

However for network sockets, by design at any one time almost all of your goroutines
are going to be blocked waiting for network IO. In a naive implementation this would 
require as many threads as goroutines, all blocked waiting on network traffic. Go’s 
integrated network poller handles this efficiently due to the cooperation between the
runtime and net packages.

In older versions of Go, the network poller was a single goroutine that was responsible
for polling for readiness notification using kqueue or epoll. The polling goroutine would
communicate back to waiting goroutines via a channel. This achieved the goal of 
avoiding a thread per syscall overhead, but used a generalised wakeup mechanism of 
channel sends. This meant the scheduler was not aware of the source or importance of
the wakeup.

In current versions of Go, the network poller has been integrated into the runtime itself.
As the runtime knows which goroutine is waiting for the socket to become ready it can
put the goroutine back on the same CPU as soon as the packet arrives, reducing latency
and increasing throughput
```

and from https://povilasv.me/go-scheduler/

```

Interesting things happen when your goroutine makes a blocking syscall. Blocking syscall
will be intercepted, if there are Gs to run, runtime will detach the thread from the P and
create a new OS thread (if idle thread doesn’t exist) to service that processor.

When a system calls resumes, the goroutine is placed back into a local run queue and
the thread will park itself (meaning thread won’t be running) and insert itself in the list
of idle threads.

If a goroutine makes a network call, runtime will do a similar action. The call will be
intercepted, but because Go has integrated network poller, which has its own thread,
it will be assigned to it.
```

Practically, golang's IO architecture envisages handling orders of magnitude larger number of sockets concurrently compared to files. Also see https://stackoverflow.com/questions/32380653/can-golang-launch-multi-thread-to-handle-networkd-io

anton-malakhov commented:
An improved libuv can do pretty much the same. It has thread pool for calling blocking fs operations, while processing of non-blocking operations can be done concurrently from regular compute/worker threads as I'm trying to prove it in my prototype [here](https://github.com/anton-malakhov/libuv):
```
[10:52:17 fxsatlin01 libuv] (multithreading/bench)$ out/Debug/run-benchmarks tcp_pump100_client
ok 1 - tcp_pump100_client
# tcp_pump10_server: 32.7 gbit/s
# tcp_pump10_client: 32.7 gbit/s
[10:55:33 fxsatlin01 libuv] (multithreading/bench)$ out/Debug/run-benchmarks tcp_pump100_threads_client
ok 1 - tcp_pump100_threads_client
# tcp_pump10_server: 52.7 gbit/s
# Thread 0x7f21deeb5f40 finished 4498174 roundtrips
# Thread 0x7f21e06b8f40 finished 4519404 roundtrips
# Thread 0x7f21df6b6f40 finished 4485044 roundtrips
# Thread 0x7f21dfeb7f40 finished 4468678 roundtrips
# Thread 0x7ffcf22bfda0 finished 4515856 roundtrips
# Unlock calls = 9539936
# tcp_pump10_client: 52.7 gbit/s
# Thread 0x7fc15e195f40 finished 1433 roundtrips
# Thread 0x7fc15e996f40 finished 11 roundtrips
# Thread 0x7fc15f197f40 finished 3020 roundtrips
# Thread 0x7fc15d994f40 finished 4 roundtrips
# Thread 0x7ffc5e7f0dc0 finished 46 roundtrips
# Unlock calls = 41407060
```
Though it is not quite stable yet. I'd like to achieve stability before creating PR against libuv. My plan is to convince libuv maintainers to introduce a notion of 'loop unlock' callback that would be enough to get closer to go-lang in benchmarks while manually implementing locks around libuv. This aligns well with libuv's design philosophy.

So, my proposal for multithreading support for Julia's I/O is to use a lock around every call to libuv (while any user-defined or switchable callback must be called without holding this lock of course). This approach is feasible with current libuv implementation (I tested it on both Linux and Windows) and will benefit from incremental modifications in libuv.

@amitmurthy, did you implement anything for I/O with your approach already?

StefanKarpinski commented:
> I'd like to achieve stability before creating PR against libuv. My plan is to convince libuv maintainers to introduce a notion of 'loop unlock' callback that would be enough to get closer to go-lang in benchmarks while manually implementing locks around libuv.

Please, please make the PR against Julia's fork of libuv first. It has some substantial differences since we have been unable to convince the upstream project to accept changes and bugfixes. If we predicate progress here on libuv accepting significant changes, I fear this will take far longer than we can afford. We will, however, continue to try to upstream all changes to the libuv project, including any you make against our fork in order to get threaded I/O working.

anton-malakhov commented:
I can do this. Though, it is not a production code anyway and it is even not on the critical path for the multithreading support. I needed this rather in order to explore a path forward so that we know how to implement I/O in Julia right now and how to improve it in the future. Since the multithreading support for libuv is a big task by itself, I wanted to create PR to libuv in order to start this discussion broadly and possibly attract somebody's help to do the actual productization of this idea while I'll be working on other parts in Julia. So, no, it is not a predicate for our work in Julia.

StefanKarpinski commented:
Great! That sounds like a solid plan. Getting I/O working using locks for 1.0 will be sufficient; post-1.0 there will definitely be others who can help out with this work (perhaps @vtjnash or @keno).

vtjnash commented:
> Please, please make the PR against Julia's fork of libuv first

I'll try to update our fork to use libuv v2.x fairly soon. That branch (upstream v2.x, aka master) is probably the best target for your work.

StefanKarpinski commented:
To clarify: our branch or their branch? Maybe a link to be 100% clear?

amitmurthy commented:
@anton-malakhov #22996 effectively takes a closure and queues it for execution on the main event loop. This is at a much higher level than intercepting individual libuv write calls and shipping them to the main event loop. I'll try and cleanup #22996 soon. I chose this model for the following reasons. 

- In the `Distributed` module, serializing a large object involves multiple `write` calls. It is more efficient to ship the entire top level serialization call to the main thread than individual `write` calls. This is not an issue for smaller objects where data is first written to an in-memory buffer and a single libuv `write` performed.

- I don't know how much of our standard library is thread-safe. Queued execution of blocks of code is simpler than putting appropriate locks in place. Also, I suspect the locking mechanisms we have need to be better integrated with the task infrastructure.

I would like to mention one more thing to keep in mind as you enhance libuv (mostly from the implementation of `Distributed` but probably equally applicable elsewhere). While a single scheduled task reads incoming requests from a socket connected to a worker, any task can write directly to the socket for outbound communication. If you envisage multiple event loops and sticky tasks, how this effects the design of socket IO can be considered.

anton-malakhov commented:
My only biggest concern with this design is that main thread will be eventually blocked in a computation which will block the entire I/O thing. Design of such a general subsystem as I/O should better not depend on particular needs of one component. I guess the shipping/queuing/executing of the serialization in single thread (for efficiency) can be done regardless of how I/O is implemented.
Making synchronization of standard library thread safe is our next unavoidable target, we cannot move forward without it anyway.
Sorry, if my earlier message made you lost in my flow of thoughts. My latest proposal is *single locked event loop*. I hesitated to do that initially but my prototype shows good enough results, thus I suggest going this way as it will work immediately and can be incrementally improved once libuv is ready.

anton-malakhov commented on base/event.jl:
```diff
@@ -2,6 +2,33 @@
+
+Condition() = ccall(:jl_condition_new, Ref{Condition}, ())
+wait(c::Condition) = ccall(:jl_task_wait, Void, (Condition,), c)
+notify(c::Condition) == ccall(:jl_task_notify, Void, (Condition,), c)
```
I might not be quite fluent in Julia yet but this `==` look suspicious.

JeffBezanson commented on base/event.jl:
```diff
@@ -2,6 +2,39 @@
+n_waiters(c::Condition) = 0 # TODO: do we need this?
+
+yield() = ccall(:jl_task_yield, Cvoid, (Cint,), 1)
+wait() = yield()
```
`wait()` should block forever (until the task is scheduled again). `yield()` should switch away from the current task, but immediately put it back in a queue so it will run again.

JeffBezanson replied:
Yes, you could call `schedule` on it, or it might just block forever and get GC'd.

vchuravy commented on src/partr.c:
```diff
@@ -0,0 +1,1055 @@
+
+/*  jl_condition_new() -- create a new Condition
+ */
+JL_DLLEXPORT jl_condition_t *jl_condition_new(void)
```
I am getting a GC seqfault on a Condition object allocated from here.

1. `JL_GC_PUSH` & `JL_GC_POP` seems to be missing here
2. Is `jl_gc_wb(cond, cond->notify)` necessary? That is a compiler warning for me.

kpamnany replied:
Fixed both.

vchuravy commented on src/julia.h:
```diff
@@ -1601,13 +1609,153 @@ typedef struct _jl_task_t {
+    jl_timing_block_t *timing_stack;
+};
+
+struct _jl_condition_t {
```
The definition of `jl_condition_t` and `jl_condition_type` have gone out of sync.

```
mutable struct Condtion
    notify::UInt8
    waitq_head::Task
    waitq_lock_owner::Int64
    waitq_lock_count::Int32
end
```

vchuravy commented on src/partr.c:
```diff
@@ -0,0 +1,1055 @@
+{
+    jl_ptls_t ptls = jl_get_ptls_states();
+
+    jl_task_t *task = (jl_task_t *)jl_gc_alloc(ptls, sizeof (jl_task_t),
```
`jl_new_struct_uninit(jl_task_type)`

vchuravy commented on src/partr.c:
```diff
@@ -0,0 +1,1055 @@
+{
+    jl_ptls_t ptls = jl_get_ptls_states();
+
+    jl_condition_t *cond = (jl_condition_t *)
```
`jl_new_struct_uninit(jl_condition_type)`

vchuravy reviewed (commented):
A couple of other things I encountered while trying to get this to compile. The biggest showstopper for me right now is that `Condtion` and `jl_condition_t` are not in sync, and I am not sure which one is the right version.

Should the  task queue have a corresponding Julia type.

Also my general feeling is that I would like to simplify this as much as possible and move things to Julia.

JeffBezanson commented on src/task.c:
```diff
@@ -721,26 +782,38 @@ void jl_init_root_task(void *stack, size_t ssize)
+    ptls->current_task->arr = NULL;
+    ptls->current_task->red = NULL;
+    ptls->current_task->red_result = jl_nothing;
+    ptls->current_task->grain_num = -1;
```
This is missing some field initializations: `args`, `mfunc`, `rargs`, `mredfunc`, etc. The gc is crashing when it hits uninitialized pointers here.

kpamnany replied:
A `jl_value_t` array like `args` (or any of the `jl_any_type` fields) should be `jl_nothing` rather than NULL, yes?

JeffBezanson replied:
Generally yes. If set to NULL, a field will be "undefined" in julia and raise an error on access. In that case the main worry is the `ninitialized` argument to `jl_new_datatype`.

JeffBezanson commented on src/task.c:
```diff
@@ -697,6 +708,56 @@ void jl_init_tasks(void)
+                                jl_any_type,
+                                jl_any_type,
+                                jl_any_type,
+                                jl_any_type,
```
These two need to be int64 and int32.

JeffBezanson commented on base/event.jl:
```diff
@@ -2,6 +2,39 @@
+
+Condition() = ccall(:jl_condition_new, Ref{Condition}, ())
+wait(c::Condition) = ccall(:jl_task_wait, Cvoid, (Ref{Condition},), Ref{c})
+notify(c::Condition, arg, all, error) = ccall(:jl_task_notify, Cvoid, (Ref{Condition},), Ref{c})
```
These should pass `c`, not `Ref{c}`. Same with `t` below. `Ref{c}` is a type, not a condition object.

JeffBezanson commented:
This is also needed:
```diff
--- a/src/staticdata.c
+++ b/src/staticdata.c
@@ -1645,6 +1645,7 @@ static void jl_init_serializer2(int for_serialize)
                      jl_globalref_type->name, jl_typeofbottom_type->name,
                      jl_string_type->name, jl_abstractstring_type->name,
                      jl_namedtuple_type, jl_namedtuple_typename,
+                     jl_condition_type, jl_condition_type->name,
 
                      jl_int32_type, jl_int64_type, jl_bool_type, jl_uint8_type,
```

JeffBezanson commented on src/task.c:
```diff
@@ -721,26 +785,44 @@ void jl_init_root_task(void *stack, size_t ssize)
+    ptls->current_task->rargs = jl_nothing;
+    ptls->current_task->mredfunc = NULL;
+    ptls->current_task->next = NULL;
+    ptls->current_task->parent = NULL;
```
This needs to point to a Task.

kpamnany replied:
`parent` can be made to point to `current_task`, but what about `next`? We need a way to determine the end of the list. Should I create a global dummy sentinel?

JeffBezanson replied:
You can use `nothing` as a sentinel, or leave it as NULL (though there is no way to set a field to NULL from julia).

JeffBezanson replied:
Currently we make the root task its own parent.

vtjnash replied:
Currently, the `parent` field is deprecated

JeffBezanson replied:
It's still referenced in the loading code, which was causing the sysimg build to fail.

kpamnany replied:
partr uses `parent` to identify the first task in the set of tasks created for a `parfor`.

JeffBezanson commented on src/task.c:
```diff
@@ -721,26 +785,44 @@ void jl_init_root_task(void *stack, size_t ssize)
+    ptls->current_task->sticky_tid = ptls->tid;
+    ptls->current_task->grain_num = -1;
+    ptls->current_task->fptr = NULL;
+    ptls->current_task->rfptr = NULL;
```
Also `ptls->current_task->cq`.

JeffBezanson commented on src/gc.c:
```diff
@@ -673,6 +673,11 @@ void jl_gc_force_mark_old(jl_ptls_t ptls, jl_value_t *v)
+#ifdef JULIA_ENABLE_PARTR
+    else if (dt == jl_condition_type) {
+        dtsz = sizeof(jl_condition_t);
+    }
```
Not necessary, since the julia struct has the same size as the C declaration.

JeffBezanson commented on src/dump.c:
```diff
@@ -784,6 +784,11 @@ static void jl_serialize_value_(jl_serializer_state *s, jl_value_t *v, int as_li
+#ifdef JULIA_ENABLE_PARTR
+    else if (jl_typeis(v, jl_condition_type)) {
+        jl_error("Condition cannot be serialized");
+    }
```
This is not really necessary; might as well allow serializing a Condition as long as it doesn't point to any non-serializable tasks.

kpamnany replied:
Right above this it says "Task cannot be serialized"?

JeffBezanson replied:
Yes, if there are tasks inside you'll get that error, but we need to be able to save empty Conditions, for example in the global semaphore in threadcall.jl.

kpamnany commented:
> jl_setjmp(ytask->ctx, 0) in jl_task_yield crashes

I fixed this by being more defensive about `ptls->current_task`.

It is now crashing because a task is somehow being scheduled with a `NULL` function pointer. Will debug.

StefanKarpinski commented:
Does all that green mean this is working now?!?

kpamnany commented:
Not quite yet, but I think we're pretty close. A task seems to be getting GC'ed when it shouldn't. There's a couple of `i`s I haven't dotted yet...

StefanKarpinski commented:
Anything anyone else can help with? Maybe @vtjnash?

vchuravy commented on src/partr.c:
```diff
@@ -689,7 +689,7 @@ JL_DLLEXPORT jl_task_t *jl_task_new(jl_value_t *_args)
     jl_ptls_t ptls = jl_get_ptls_states();
 
-    jl_task_t *task = (jl_task_t *)jl_new_struct_uninit(jl_task_type);
+    jl_task_t *task = (jl_task_t *)jl_gc_alloc(ptls, sizeof (jl_task_t), jl_task_type);
```
Why are you using `jl_gc_alloc` directly? 

Is there a size difference between `jl_datatype_size(jl_task_type)`  and `sizeof(jl_task_t)`?

Keno replied:
Sadly, yes. `jl_task_t` has a bunch of hidden fields.

vchuravy commented on base/event.jl:
```diff
@@ -5,33 +5,114 @@
 Condition() = ccall(:jl_condition_new, Ref{Condition}, ())
+
+"""
+    wait([x])
```
I think we want to attach this docstring (since it is so general) to `function wait end` and not `wait(condition)`.
IIUC docstrings get accumulated over methods and shown when asked for the generic function one, but you can also ask for a specific docstring once you have a method

kpamnany commented:
A lot got done this week. I'll update the first post sometime next week most likely.

JeffBezanson commented:
Noting that the work on stack management is in #13099. These PRs will need to be combined eventually.

tveldhui commented:
Is this branch expected to build on OS X?  On my mac (OS X 10.12.6) it crashes on startup if JULIA_NUM_THREADS > 1.  lldb shows no particular pattern other than EXC_BAD_ACCESS with obviously bad pointers, like address=0x8.

```
(lldb) run
Process 49772 launched: '/Users/tveldhui/RelationalAI/julia/usr/bin/julia' (x86_64)
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.5129 (2018-05-18 18:16 UTC)
 _/ |\__'_|_|_|\__'_|  |  kp/partr/362f99158e* (fork: 23 commits, 9 days)
|__/                   |  x86_64-apple-darwin16.7.0


signal (11): Segmentation fault: 11
in expression starting at no file:0
uv__update_time at /Users/tveldhui/RelationalAI/julia/deps/srccache/libuv-be317349252699670131395f125c3861d793ca86/src/unix/internal.h:274
uv__hrtime at /Users/tveldhui/RelationalAI/julia/deps/srccache/libuv-be317349252699670131395f125c3861d793ca86/src/unix/darwin.c:59
uv__update_time at /Users/tveldhui/RelationalAI/julia/deps/srccache/libuv-be317349252699670131395f125c3861d793ca86/src/unix/internal.h:274
uv__io_poll at /Users/tveldhui/RelationalAI/julia/deps/srccache/libuv-be317349252699670131395f125c3861d793ca86/src/unix/kqueue.c:175
Allocations: 461756 (Pool: 461623; Big: 133); GC: 0
Process 49772 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x2d0)
    frame #0: 0x00000001001f2c2a libjulia.dylib`uv__update_time(loop=0x0000000000000000) at internal.h:274
   271 	UV_UNUSED(static void uv__update_time(uv_loop_t* loop)) {
   272 	  /* Use a fast time source if available.  We only need millisecond precision.
   273 	   */
-> 274 	  loop->time = uv__hrtime(UV_CLOCK_FAST) / 1000000;
   275 	}
   276 	

```
thanks!

kpamnany commented:
Rebased on master and cleaned up a little. Updated the first post with the current status.

@tveldhui: as the first post now makes clear, don't expect multiple threads to work right now. Since threads are now tightly integrated with tasks, the impact of parallelism on the standard libraries (and other packages that use tasks) is now far greater -- more things need to become thread-safe.

ViralBShah commented:
If this by and large works, and is anyways guarded behind a flag, are we at a point where we should consider merging it and do more thread-safe things subsequently with new issues/PRs?

oxinabox commented on base/channels.jl:
```diff
@@ -2,6 +2,56 @@
 
 abstract type AbstractChannel end
 
+if JULIA_PARTR
```
should this be `@static if`?

StefanKarpinski replied:
they're equivalent at the top-level

oxinabox replied:
Good to know, thanks.
There are a few places further down that are not at top-level that should be `@static if` then,
I've now marked them

oxinabox commented on base/channels.jl:
```diff
@@ -2,6 +2,56 @@
+    lock::SpinLock
+
+    # Used when sz_max == 0, i.e., an unbuffered channel.
+    waiters::Atomic{Int}
```
What is used when `sz_max == 0` ? `waiters`? Or the whole block of things after that comment?
(or the lock above)

Should waiters be `n_waiters` or `num_waiters` or `nwaiting` ?
I was confusers seeing it as an `Int` and not as a `Vector{Task}` like `takers` and `putters`
I had to read ahead to workout that the `Int` was a count.
(and idk, not some kind of ID or index into an array or just a typo)

oxinabox commented on base/channels.jl:
```diff
@@ -275,12 +368,41 @@ function put_unbuffered(c::Channel, v)
+remove the item. `fetch` is unsupported on an unbuffered (0-size) channel.
+"""
+function fetch(c::Channel)
+    c.sz_max == 0 && throw(ErrorException("`fetch` is not supported on an unbuffered Channel"))
```
It would be good to change this to something other than `ErrorException` 
Since that type gives no useful information for catching purposes.
It is thrown a lot by user level (i.e. not pkg level) because it is the type of `error("Oh Noooo")`.

oxinabox commented on base/channels.jl:
```diff
@@ -304,14 +427,54 @@ For unbuffered channels, blocks until a [`put!`](@ref) is performed by a differe
+        unlock(c.lock)
+        rethrow(ex)
+    end
+end
```
The locks/unlocks in this section could be written in `lock(func)` form.
In 0.6 I found this had a small overhead over writing `lock` then `unlock` ,
but in theory it should just inline the anon func out of existence.

oxinabox replied:
As a side point, I like that the PATR code for this one is a lot clearer than the nonPATR code :-D

oxinabox commented on base/channels.jl:
```diff
@@ -304,14 +427,54 @@ For unbuffered channels, blocks until a [`put!`](@ref) is performed by a differe
+    unlock(c.lock)
+    notify(c.cond_put, nothing, false, false)
+    try
+        return wait()::T
```
is it worth commenting this line, to explain that `wait` actually returns a value?
I guess it is obvious since otherwise this line would be nonsense

oxinabox commented on base/channels.jl:
```diff
@@ -344,7 +511,17 @@ For unbuffered channels returns `true` if there are tasks waiting
+    lock(c.lock)
+    n = isbuffered(c) ? length(c.data) : length(c.putters)
+    unlock(c.lock)
+    return n
```
I think a good one to use `lock(func, lk)` form with.
```
n_avail(c::Channel) = lock(c.lock) do
     isbuffered(c) ? length(c.data) : length(c.putters)
end
```

oxinabox commented on base/event.jl:
```diff
@@ -2,6 +2,121 @@
+"""
+    yield()
+
+Allow the scheduler to use the thread running the current task to run a higher priority task,
```
This is the only time in this whole PR that the word `prority` is used.
Is priority a real thing? If so how is it determined?

oxinabox commented on base/event.jl:
```diff
@@ -2,6 +2,121 @@
+yield() = ccall(:jl_task_yield, Any, (Cint,), 1)
+yield(t::Task, @nospecialize x = nothing) = (schedule(t, x); yield())
+yieldto(t::Task, @nospecialize x = nothing) = yield(t, x)
+try_yieldto(undo, reftask::Ref{Task}) = yield(t, x)
```
there is an error in `try_yeildto` the variables `t` and `x` are not defined

oxinabox commented on base/stream.jl:
```diff
@@ -278,8 +278,14 @@ function wait_readbyte(x::LibuvStream, c::UInt8)
+            if isempty(x.readnotify)
+                stop_reading(x) # stop reading iff there are currently no other read clients of the stream
+            end
+        else
```
This `else` and some others are missing the comments saying `## ! JULIA_PARTR`.
I had got used to using those to navigate, they are helpful.

oxinabox commented:
this is a huge and awesome PR. I am very looking forward to it.
I only got through the julia stuff in my reading (and I didn't look really hard for race conditions yet either).
Please ignore any of my comments that you feel are wrong or pedantic 

I am looking forward to this.

oxinabox commented on base/stream.jl:
```diff
@@ -278,8 +278,14 @@ function wait_readbyte(x::LibuvStream, c::UInt8)
     finally
-        if isempty(x.readnotify.waitq)
-            stop_reading(x) # stop reading iff there are currently no other read clients of the stream
+        if JULIA_PARTR
```
This should be `@static if` I think

oxinabox commented on base/stream.jl:
```diff
@@ -301,8 +307,14 @@ function wait_readnb(x::LibuvStream, nb::Int)
     finally
-        if isempty(x.readnotify.waitq)
-            stop_reading(x) # stop reading iff there are currently no other read clients of the stream
+        if JULIA_PARTR
```
This should be `@static if`  I think?

kpamnany commented:
@oxinabox: thanks for the review! A number of things in `Channel`s still confuse me -- I've only been working on getting things working. @amitmurthy: I think you're familiar with channels; perhaps you could have a look at some of Lyndon's comments, or point someone else at them? I think the whole implementation could use some refactoring rather than the patch job I've done.

I'll address the comments I do know what to do about in the next commit.

Let me reiterate that the full focus right now is on getting this working (Julia's test suite passes) with a single thread. Getting multi-threaded functionality working properly will be left to future PRs; this one is too large as it is!

kpamnany commented:
Okay, finally got the channels tests to pass! I'll run the Distributed tests next.

The tests for `AsyncCondition` depend on task execution order. This is not meaningful for parallel tasks and I think they should be discarded/rewritten. Nonetheless, many work with `JULIA_PARTR` on, but strangely don't work with it turned off. I haven't really looked into why this is but if anyone can figure out what's wrong there, that'd be great.

Attn: @JeffBezanson, @vtjnash, @vchuravy.

vchuravy commented on base/channels.jl:
```diff
@@ -104,13 +157,12 @@ function Channel(func::Function; ctype=Any, csize=0, taskref=nothing)
     task = Task(() -> func(chnl))
     bind(chnl, task)
-    yield(task) # immediately start it
+    schedule(task)
```
Should we deprecate `yield(task)` to `schedule(task)` on 0.7, immediately?

kpamnany replied:
I'd like to. And, I'm even more interested in deprecating `yieldto()`! But I haven't won that argument yet...

vchuravy commented on test/channels.jl:
```diff
@@ -127,7 +128,7 @@ using Distributed
     # channeled_tasks
     for T in [Any, Int]
-        chnls, tasks = Base.channeled_tasks(2, (c1,c2)->(@assert take!(c1)==1; put!(c2,2)); ctypes=[T,T], csizes=[N,N])
+        chnls, tasks = Base.channeled_tasks(2, (c1,c2)->(@assert take!(c1)==1; put!(c2,2); sleep(1.0)); ctypes=[T,T], csizes=[N,N])
```
are the `sleep`'s here necessary?

anton-malakhov replied:
good question. If this sleep is necessary to get another task executed, why not use direct means like conditional variable (if it fits the test's paradigm)

kpamnany replied:
That's why I added the comment above this testset that these tests should be redesigned -- for some reason, all these tests rely on an understanding of how tasks are scheduled (or rather how tasks _were_ scheduled). With the new runtime, such assumptions no longer hold.

vchuravy commented on src/partr.c:
```diff
@@ -0,0 +1,1166 @@
+        init_task(task, _args);
+        task->start = start;
+        task->end = end;
+        task->parent = parent;
```
this should probably be `task->parent = prev` or should it really point to the topmost parent?
I thought this was a double-linked-list.

kpamnany replied:
It's not a doubly linked list. There's a tree structure derived from the arriver (`task->arr`) for synchronization. The `parent` represents the set of tasks and is what we return to the caller. It's used by other tasks to record the final reduction result (when there's a reduction requested) and for a couple of other things (see `sync_grains()` etc.).

kpamnany commented:
I should note that the problem I was seeing with `JULIA_PARTR` turned off is now fixed -- this branch now behaves just like master when `JULIA_PARTR` is not defined.

The current issue holding us up seems GC-related; Valentin is helping with it but any other help is of course welcome.

Also: can someone help me understand why Documenter is unable to find docs for `Base.yield` etc. when the documentation is clearly there (in `base/event.jl`)?

orenbenkiki commented:
For planning purposes - is there an expected time frame for merging this?

I have a Python/C++/Pandas/Numpy project I'm considering migrating to Julia. I definitely need multi-threading support, and having a multi-threaded parallel task scheduler would solve a lot of problems for me.

I tried to roll my own scheduler as a temporary measure (https://github.com/orenbenkiki/snarl) and discovered it is currently impossible to pause/notify a thread in Julia without busy-waiting, and that almost everything isn't thread safe (including `sleep`, which makes busy-waiting even more painful).

Would it be reasonable to expect the PR to be merged (and included in a point release) within, say, this year?

vchuravy commented:
> For planning purposes - is there an expected time frame for merging this?

My careful answer would be "probably this year", but it not certain in which form and when.
Most likely we will merge this first under an experimental flag and take it from there.

orenbenkiki commented:
>    For planning purposes - is there an expected time frame for merging this?

> > My careful answer would be "probably this year", but it is not certain in which form and when.

Thanks. I'll keep tracking this then. Other than the multi-threading not working, I found coding in Julia to be pretty pleasant so far. Certainly a better experience than mixing C++ and Python together :-)

vchuravy commented:
@kpamnany if you squash/rebase this feel free to drop my ASAN/valgrind commits since those were debugging only and I will add it back separately.

Red-Portal commented:
Do you guys think future continuation API in the [fashion of C++](https://en.cppreference.com/w/cpp/experimental/future/then) would be necessary?
This would make possible to chain computational dependencies without requiring them to return task objects.
I think something using the pipe operator would look nice
For example, for a matrix-vector multiplication problem,

```julia
A = randn(N,N)
x = randn(N)
y = zeros(N)
for i=1:N
   y[i] = @schedule (@task mult(A[i,:], x) |> sum)
end
```

JeffBezanson commented:
I think I see; so would `a |> b` mean something like `b(fetch(@async a))` ?

Red-Portal commented:
@JeffBezanson Yes that is correct.

JeffBezanson commented:
The regression from using `jl_switchto` seems to be related to this condition in that function:
```
    if (//t->state == done_sym || t->state == failed_sym ||
        (t->started && t->stkbuf == NULL)) {
        ptls->current_task->exception = t->exception;
        ptls->current_task->result = t->result;
        return;
    }
```
Note I commented out the first part of the condition; only the second part causes the hang.

JeffBezanson commented:
Got it: the `started` field was not initialized.

vchuravy commented:
Started a couple of buildbots with PARTR enabled off 0a412c2b9d993ca032dcefea937e319b5c766dcf

- linux64: https://build.julialang.org/#/builders/53/builds/94
- linux32: https://build.julialang.org/#/builders/35/builds/90
- win64: https://build.julialang.org/#/builders/64/builds/90

JeffBezanson commented:
The failure in Distributed is due to the code there making assumptions about the order of task execution. Each process assumes that the first message it gets from another process is special in a certain way, and that order can be violated in the following way:
- process 3 launches a task to send its welcome message to process 2
- process 1 tells process 3 to send a message to process 2
- process 3 sends the message from process 1 to process 2 (too soon)
- process 3 sends its welcome message (too late)

JeffBezanson commented:
Remaining issues are FileWatching and Sockets tests hanging.

JeffBezanson commented:
Big milestone: passes tests with partr enabled on mac and freebsd! :tada: 
Hanging on windows in generate_precompile.jl.

kpamnany commented:
> The failure in Distributed is due to the code there making assumptions about the order of task execution.

I have no idea how you traced this from a zeroed packet! 🙇

kpamnany commented:
- Linux 32: freezing in `generate_precompile`. :(
- Linux 64: all tests pass, some error in `makedocs`.
- OSX 64: passes. :)
- Win *: haven't finished yet? But seem to be frozen in `generate_precompile`. :(
- FreeBSD: passes. :)

Cc: @JeffBezanson

kpamnany commented:
Linux 32 is also freezing in `generate_precompile`. Sadly, all these are platforms I have no access to.

fredrikekre commented:
> what looks like a misplaced newline (?) in doctest output.

Here is a MWE for the doctest failure:
```
function doctest()
    default_stdout = stdout
    default_stderr = stderr

    pipe = Pipe()
    Base.link_pipe!(pipe; reader_supports_async = true, writer_supports_async = true)
    redirect_stdout(pipe.in)
    redirect_stderr(pipe.in)

    output = UInt8[]

    exs = Meta.parse.(["chnl = Channel(c->print(take!(c)); taskref=Ref{Task}())",
                       "put!(chnl, \"Hello\")"])

    sandbox = Module()

    try
        for ex in exs
            Core.eval(sandbox, ex)
        end
    catch err

    finally
        # Force at least a single write to `pipe`, otherwise `readavailable` blocks.
        println()

        redirect_stdout(default_stdout)
        redirect_stderr(default_stderr)
        append!(output, readavailable(pipe))
        close(pipe)
    end
    String(output)
end
```
which return `Hello\n` on master and `\nHello` on this branch.

JeffBezanson commented:
Excellent, that is very helpful. It looks like the general problem is that the doctest framework has no way to know when tasks spawned by doc examples are done writing all their output. I think we're missing a `yield` in `put!`, which might be enough to fix this, but the general problem is harder. Only thing that comes to mind is putting a bunch of `yield()` calls before the `println()` :-P

ViralBShah commented:
> Linux 32 is also freezing in `generate_precompile`. Sadly, all these are platforms I have no access to.

The easiest thing is to set up a virtualbox image. Perhaps @Keno may have something already on anubis.

fredrikekre commented:
> I think we're missing a yield in put!, which might be enough to fix this

Looks like it did? Docs build succesfully on Travis.

vtjnash commented:
Almost any use of `readavailable` is inherently race-y, which is why this is now failing. We could rewrite that code correctly, then that problem would go away.

KristofferC commented:
Should it write some sentinel to the end and use `readuntil`?

c42f commented on src/task.c:
```diff
@@ -512,12 +532,57 @@ void jl_init_tasks(void) JL_GC_DISABLED
+                                jl_any_type,
+                                jl_any_type,
+                                jl_any_type,
+                                jl_int64_type,
```
This `jl_int64_type` looks fishy. It's a part of `jl_mutex_t` which contains `volatile unsigned long owner` which is 32 bits on linux32.

c42f commented on src/partr.c:
```diff
@@ -0,0 +1,1149 @@
+{
+    heap_p = heap_c * jl_n_threads;
+    heaps = (taskheap_t *)calloc(heap_p, sizeof(taskheap_t));
+    for (int16_t i = 0;  i < heap_p;  ++i) {
```
Why the `int16_t` for the iterator here? Perhaps a native width `int` (or `size_t`) would make more sense?

Same comment applies to various iterators within partr.c - some are `int16_t`, some `int`, there's even an `int64_t`.

c42f replied:
Right. I think it should mainly be necessary to match signed vs unsigned in comparisons (not so much bit width) to avoid surprises and related warnings arising from the likes of

```C
#include <stdio.h>
int main() {
    if (-1 > (size_t)0)
        printf("Surprise!\n");
    return 0;
}
```

I mainly noticed this because I was having a read through for where 32 and 64 bit versions might differ.

c42f commented on src/task.c:
```diff
@@ -512,12 +532,57 @@ void jl_init_tasks(void) JL_GC_DISABLED
+    jl_condition_type = (jl_datatype_t*)
+        jl_new_datatype(jl_symbol("Condition"), NULL, jl_any_type, jl_emptysvec,
+                        jl_perm_symsvec(3, "head", "lock_owner", "lock_count"),
+                        jl_svec(3, jl_task_type, jl_int64_type, jl_int32_type),
```
Well, I just went ahead and fixed this directly to see what would happen in CI. Hope that's ok!

kpamnany commented:
Looks like the `generate_precompile` step now passes fine. 🎆 

Now, all platforms seem to get to the same point:
```
┌ Warning: TerminalMenus: Unable to enter raw mode: ArgumentError("stream is closed or unusable")
└ @ REPL.TerminalMenus /usr/home/julia/julia-fbsd-buildbot/worker/freebsdci/build/usr/share/julia/stdlib/v1.1/REPL/src/TerminalMenus/util.jl:21
```

Running `./julia test/runtests.jl REPL` interactively (on OS-X) gives me:
```
REPL: Test Failed at /Users/kpamnany/julia/usr/share/julia/stdlib/v1.1/REPL/test/replcompletions.jl:107
  Expression: count(isequal("REPL"), c) == 1
   Evaluated: 2 == 1                 
```
Which seems unrelated? 😕

KristofferC commented:
> ┌ Warning: TerminalMenus: Unable to enter raw mode: ArgumentError("stream is closed or unusable")

Is a warning that is always printed (we should fix that) but it is unrelated to the PR.

The actual problem is

```
No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

which means that one of the tests likely got stuck.

c42f commented:
Shall I merge #29791 in here?

kpamnany commented:
Looks like a big patch, and this is a big patch too with some time pressure to merge. @JeffBezanson can make the call.

Keno commented:
@kpamnany note that this branch has conflicts with master. @c42f's PR resolves those conflicts.

kpamnany commented:
Ah, I see now, it is already on master so it has to be merged. It'd be best if we could get `runtests` to complete successfully on this branch first before adding new code to fix though. Unless the new code can help?

c42f commented:
The new code is likely to help in the circumstance that you're loosing stack traces of exceptions thrown in tasks due to context switching. Other than that, it only resolves the conflicts with master.

JeffBezanson commented:
Ok I think I found the next problem. wait/isready/n_avail depend on the length of `c.putters`, but the partr code doesn't use that array any more. Will change it to use the state of `cond_put`.

JeffBezanson commented:
Now other tests pass but there is a mystery crash in the embedding test. :rage:

c42f commented on src/task.c:
```diff
@@ -220,6 +220,12 @@ JL_DLLEXPORT void *jl_task_stack_buffer(jl_task_t *task, size_t *size, int *tid)
+static void record_backtrace(jl_ptls_t ptls) JL_NOTSAFEPOINT
+{
+    // storing bt_size in ptls ensures roots in bt_data will be found
+    ptls->bt_size = rec_backtrace(ptls->bt_data, JL_MAX_BT_SIZE);
```
I agree this is nice to have factored out (especially for having a place to put the note about rooting). I only removed it for symmetry with the equivalent code which had removed it in `partr.c`. We should probably just call `record_backtrace` there as well.

kpamnany commented:
Such good progress! The Win* failures are mystifying -- I see a lot of `fatal: cannot change to '/cygdrive/c/projects/julia': No such file or directory` messages?

We're still going to turn `JULIA_PARTR` off before merge I would think?

vtjnash committed:
revert invalid changes

JeffBezanson commented:
This branch still ought to work on its own, at least with 1 thread.

JeffBezanson committed:
try enabling again

vtjnash committed:
fix behavior divergence with non-partr build

JeffBezanson commented:
win32: generate_precompile hanging
win64: a node 1 test is hanging, either precompile, SharedArrays, or Distributed

kpamnany commented:
Refactored into https://github.com/JuliaLang/julia/pull/30806, https://github.com/JuliaLang/julia/pull/30838 and https://github.com/JuliaLang/julia/pull/31398.
